File: Algorithms/p240_Search_a_2D_Matrix_II_1.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(matrix.size()==0) return 0;
        int m = matrix.size();
        int n = matrix[0].size();
        bool h = (n>m);
        int bound1 = h ? n : m;
        int bound2 = h ? m : n;
        for(int k=0,end=bound1;k<bound2;++k){
            int low = 0;
            int high = end;
            while(low<high){
                int mid = (low+high)/2;
                int v = h ? matrix[k][mid] : matrix[mid][k];
                if(v<target){
                    low = mid+1;
                }else if(v>target){
                    high = mid;
                }else{
                    return 1;
                }
            }
            end = low;
        }
        return 0;
    }
};

File: Algorithms/p240_Search_a_2D_Matrix_II/p240_Search_a_2D_Matrix_II_1.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if not matrix:
            return False
        m = len(matrix)
        n = len(matrix[0])
        h = (n>m)
        bound1 = n if h else m
        bound2 = m if h else n
        end=bound1
        for k in range(0,bound2):
            low = 0
            high = end
            while low<high:
                mid = (low+high)//2
                v = matrix[k][mid] if h else matrix[mid][k]
                if v<target:
                    low = mid+1
                elif v>target:
                    high = mid
                else:
                    return True
            end = low
        return False
